File: src/test/java/com/github/nigelzor/hogs/ShiftMatrixTest.java
package com.github.nigelzor.hogs;

import static org.junit.Assert.*;

import org.junit.Test;

public class ShiftMatrixTest {

	@Test
	public void testShiftingOneRight() {
		ShiftMatrix<String> moving = fill(new ShiftMatrix<String>(2, 2));
		moving.set(1, 1, null);
		moving.shift(1, 0);

		ShiftMatrix<String> expected = fill(new ShiftMatrix<String>(2, 2));
		expected.set(1, 0, null);
		expected.set(1, 1, "C");

		assertEquals(expected, moving);
	}

	@Test
	public void testShiftingOneLeft() {
		ShiftMatrix<String> moving = fill(new ShiftMatrix<String>(2, 2));
		moving.set(0, 0, null);
		moving.shift(0, 1);

		ShiftMatrix<String> expected = fill(new ShiftMatrix<String>(2, 2));
		expected.set(0, 0, "B");
		expected.set(0, 1, null);

		assertEquals(expected, moving);
	}

	@Test
	public void testShiftingTwoLeft() {
		ShiftMatrix<String> moving = fill(new ShiftMatrix<String>(1, 3));
		moving.set(0, 0, null);
		moving.shift(0, 2);

		ShiftMatrix<String> expected = new ShiftMatrix<String>(1, 3);
		expected.set(0, 0, "B");
		expected.set(0, 1, "C");
		expected.set(0, 2, null);

		assertEquals(expected, moving);
	}

	@Test
	public void testShiftingPartOfRowLeft() {
		ShiftMatrix<String> moving = fill(new ShiftMatrix<String>(1, 3));
		moving.set(0, 0, null);
		moving.shift(0, 1);

		ShiftMatrix<String> expected = new ShiftMatrix<String>(1, 3);
		expected.set(0, 0, "B");
		expected.set(0, 1, null);
		expected.set(0, 2, "C");

		assertEquals(expected, moving);
	}

	@Test
	public void testShiftingOneDown() {
		ShiftMatrix<String> moving = fill(new ShiftMatrix<String>(2, 2));
		moving.set(1, 1, null);
		moving.shift(0, 1);

		ShiftMatrix<String> expected = fill(new ShiftMatrix<String>(2, 2));
		expected.set(0, 1, null);
		expected.set(1, 1, "B");

		assertEquals(expected, moving);
	}

	@Test
	public void testShiftingOneUp() {
		ShiftMatrix<String> moving = fill(new ShiftMatrix<String>(2, 2));
		moving.set(0, 0, null);
		moving.shift(1, 0);

		ShiftMatrix<String> expected = fill(new ShiftMatrix<String>(2, 2));
		expected.set(1, 0, null);
		expected.set(0, 0, "C");

		assertEquals(expected, moving);
	}

	@Test
	public void testShiftingTwoUp() {
		ShiftMatrix<String> moving = fill(new ShiftMatrix<String>(3, 1));
		moving.set(0, 0, null);
		moving.shift(2, 0);

		ShiftMatrix<String> expected = new ShiftMatrix<String>(3, 1);
		expected.set(0, 0, "B");
		expected.set(1, 0, "C");
		expected.set(2, 0, null);

		assertEquals(expected, moving);
	}

	@Test
	public void testShiftingPartOfRowUp() {
		ShiftMatrix<String> moving = fill(new ShiftMatrix<String>(3, 1));
		moving.set(0, 0, null);
		moving.shift(1, 0);

		ShiftMatrix<String> expected = new ShiftMatrix<String>(3, 1);
		expected.set(0, 0, "B");
		expected.set(1, 0, null);
		expected.set(2, 0, "C");

		assertEquals(expected, moving);
	}

	// A B
	// C D
	private static ShiftMatrix<String> fill(ShiftMatrix<String> matrix) {
		char value = 'A';
		for (int row = 0; row < matrix.getRows(); row++) {
			for (int col = 0; col < matrix.getCols(); col++) {
				matrix.set(row, col, String.valueOf(value++));
			}
		}
		return matrix;
	}

}

File: src/test/kotlin/com/github/nigelzor/hogs/ShiftMatrixTest.kt
package com.github.nigelzor.hogs

import kotlin.test.assertEquals
import org.junit.Test

class ShiftMatrixTest {

	companion object {
		private fun fill(matrix: ShiftMatrix<String>): ShiftMatrix<String> {
			var value: Char = 'A'
			for (row in 0 until matrix.rows) {
				for (col in 0 until matrix.cols) {
					matrix[row, col] = String(charArrayOf(value++))
				}
			}
			return matrix
		}
	}

	@Test fun testShiftingOneRight() {
		val moving = fill(ShiftMatrix.empty<String>(2, 2))
		moving[1, 1] = null
		moving.shift(1, 0)
		val expected = fill(ShiftMatrix.empty<String>(2, 2))
		expected[1, 0] = null
		expected[1, 1] = "C"
		assertEquals(expected, moving)
	}

	@Test fun testShiftingOneLeft() {
		val moving = fill(ShiftMatrix.empty<String>(2, 2))
		moving[0, 0] = null
		moving.shift(0, 1)
		val expected = fill(ShiftMatrix.empty<String>(2, 2))
		expected[0, 0] = "B"
		expected[0, 1] = null
		assertEquals(expected, moving)
	}

	@Test fun testShiftingTwoLeft() {
		val moving = fill(ShiftMatrix.empty<String>(1, 3))
		moving[0, 0] = null
		moving.shift(0, 2)
		val expected = ShiftMatrix.empty<String>(1, 3)
		expected[0, 0] = "B"
		expected[0, 1] = "C"
		expected[0, 2] = null
		assertEquals(expected, moving)
	}

	@Test fun testShiftingPartOfRowLeft() {
		val moving = fill(ShiftMatrix.empty<String>(1, 3))
		moving[0, 0] = null
		moving.shift(0, 1)
		val expected = ShiftMatrix.empty<String>(1, 3)
		expected[0, 0] = "B"
		expected[0, 1] = null
		expected[0, 2] = "C"
		assertEquals(expected, moving)
	}

	@Test fun testShiftingOneDown() {
		val moving = fill(ShiftMatrix.empty<String>(2, 2))
		moving[1, 1] = null
		moving.shift(0, 1)
		val expected = fill(ShiftMatrix.empty<String>(2, 2))
		expected[0, 1] = null
		expected[1, 1] = "B"
		assertEquals(expected, moving)
	}

	@Test fun testShiftingOneUp() {
		val moving = fill(ShiftMatrix.empty<String>(2, 2))
		moving[0, 0] = null
		moving.shift(1, 0)
		val expected = fill(ShiftMatrix.empty<String>(2, 2))
		expected[1, 0] = null
		expected[0, 0] = "C"
		assertEquals(expected, moving)
	}

	@Test fun testShiftingTwoUp() {
		val moving = fill(ShiftMatrix.empty<String>(3, 1))
		moving[0, 0] = null
		moving.shift(2, 0)
		val expected = ShiftMatrix.empty<String>(3, 1)
		expected[0, 0] = "B"
		expected[1, 0] = "C"
		expected[2, 0] = null
		assertEquals(expected, moving)
	}

	@Test fun testShiftingPartOfRowUp() {
		val moving = fill(ShiftMatrix.empty<String>(3, 1))
		moving[0, 0] = null
		moving.shift(1, 0)
		val expected = ShiftMatrix.empty<String>(3, 1)
		expected[0, 0] = "B"
		expected[1, 0] = null
		expected[2, 0] = "C"
		assertEquals(expected, moving)
	}

}
